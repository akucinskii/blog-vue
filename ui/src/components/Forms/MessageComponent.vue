<script setup lang="ts">
import { computed } from "vue";
const props = defineProps({
  text: {
    type: String,
    required: false,
    default: "",
  },
  type: {
    type: String,
    required: false,
    default: "success",
  },
});

const type = computed(() => {
  if (props.type === "error") {
    return "error";
  } else {
    return "success";
  }
});
</script>

<template>
  <v-row class="my-4">
    <v-alert id="message" :type="type" :color="type" icon="mdi-alert-circle-outline" close text-color="white" my="2">
      {{ props.text }}
    </v-alert>
  </v-row>
</template>
