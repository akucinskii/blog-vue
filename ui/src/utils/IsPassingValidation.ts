import { RuleI } from "../types/types";

export const isPassingValidation = (tested: string, rules: RuleI[]) =>
  rules.every((rule: RuleI) => rule(tested) === true);
