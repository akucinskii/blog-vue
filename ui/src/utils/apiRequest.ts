import axios from "axios";

const BASE_API_URL = process.env.API_URL;

export enum HTTP_METHODS {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  DELETE = "DELETE",
  PATCH = "PATCH",
}

axios.defaults.baseURL = process.env.API_URL;

const apiRequest = async (method: HTTP_METHODS, url: string, data = {}, headers = {}) => {
  const token = localStorage.getItem("token");

  axios.defaults.headers.common["Authorization"] = "Bearer " + token;

  const response = await axios({
    method,
    url: `${BASE_API_URL}${url}`,
    data,
    headers,
  });

  return response;
};

export default apiRequest;
