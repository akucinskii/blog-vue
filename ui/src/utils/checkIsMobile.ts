const checkIsMobile = (): boolean => {
  return window.innerWidth < 1264;
};

export default checkIsMobile;
