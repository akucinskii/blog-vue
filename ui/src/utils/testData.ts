const testPosts = [
  {
    id: "313553ff-784f-497f-adfc-7e549bebab66",
    title:
      "POST TESTOWY asm.nfnasnfas.fa.sfn,acsaf;;;;;;;;;;;;jslfsafsjafjsajkfsjfsajfsafsajlfjlksajklfasjklafsjklfsajklaflkjslksfnsaf,mnfsa,fsannm",
    date: "2037-01-01",
    text: "It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using 'Content here, content here', making it look like readable English. Many desktop publishing packages and web page editors now use Lorem Ipsum as their default model text, and a search for 'lorem ipsum' will uncover many web sites still in their infancy. Various versions have evolved over the years, sometimes by accident, sometimes on purpose (injected humour and the like).",
    imageUrl:
      "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
    author: {
      username: "Robert",
      avatar:
        "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
      description: "test description",
    },
    category: {
      id: "1",
      name: "Technology",
    },
    disabled: false,
  },
  {
    id: "313553ff-784f-497f-adfc-7e549rerab66",
    title: "sfakfasjlasdasdasndaskdjaskjsj",
    date: "2020-01-01",
    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi e",
    imageUrl:
      "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
    author: {
      username: "Wojciech",
      avatar:
        "https://images.unsplash.com/photo-1495753379358-73c76ccd644b?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1539&q=80",
      description: "test description",
    },
    category: {
      id: "2",
      name: "Dieta",
    },
    disabled: false,
  },
  {
    id: "fsafasd",
    title: "Najlepsza dieta w okolicy. Zobacz teraz!!!",
    date: "2020-01-01",
    text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi euismod nisi. Nullam euismod, nisi eu consectetur euismod, nisl nisi consectetur nisi, eu consectetur nisi nisi e",
    imageUrl:
      "https://images.unsplash.com/photo-1529565214304-a882ebc5a8e6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2070&q=80",
    author: {
      username: "Test author",
      avatar:
        "https://images.unsplash.com/photo-1633621412960-6df85eff8c85?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=627&q=80",
      description: "test description",
    },
    category: {
      id: "1",
      name: "Technology",
    },
    disabled: false,
  },
  {
    id: "fsafasd",
    title: "This post shouldnt be visible",
    date: "2020-01-01",
    text: "This post shouldnt be visible",
    imageUrl:
      "https://images.unsplash.com/photo-1529565214304-a882ebc5a8e6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2070&q=80",
    author: {
      username: "Test author",
      avatar:
        "https://images.unsplash.com/photo-1633621412960-6df85eff8c85?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=627&q=80",
      description: "test description",
    },
    category: {
      id: "2",
      name: "Dieta",
    },
    disabled: true,
  },
];

export const testCategories = [
  {
    id: "1",
    name: "Technology",
  },
  {
    id: "2",
    name: "Dieta",
  },
  {
    id: "3",
    name: "Entertainment",
  },
];

export const testUser = {
  id: "1",
  username: "TestUser",
  password: "Password123.",
  avatar_url: "testAvatar.jpg",
  description: "test description",
  name: "Test name",
  surname: "Test surname",
  is_admin: true,
};

export default testPosts;
