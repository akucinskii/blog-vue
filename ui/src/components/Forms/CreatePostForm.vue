<script setup lang="ts">
import { AxiosError } from "axios";
import { ref } from "vue";
import { useCategoriesStore } from "../../stores/categories";
import { MessageInterface, ResponseInterface } from "../../types/types";
import apiRequest, { HTTP_METHODS } from "../../utils/apiRequest";
import MessageComponent from "./MessageComponent.vue";

const categoriesStore = useCategoriesStore();

const message = ref<MessageInterface>({
  text: "",
  type: "success",
});

const postForm = ref({
  title: "",
  text: "",
  category_id: "",
  image_url: "",
  big_image_url: "robercik.jpg", //api needs this value, but it's not used
  disabled: false,
});

const handleCreatePost = async () => {
  try {
    await apiRequest(HTTP_METHODS.POST, "/posts", postForm.value);
    postForm.value = {
      title: "",
      text: "",
      category_id: "",
      image_url: "",
      big_image_url: "",
      disabled: false,
    };

    message.value = {
      text: "Post created successfully",
      type: "success",
    };
  } catch (error) {
    const err = error as AxiosError<ResponseInterface>;

    message.value.type = "error";
    message.value.text = err.response?.data.detail ?? "Something went wrong";
  }
};

defineExpose({
  message,
  postForm,
  handleCreatePost,
});
</script>

<template>
  <v-sheet elevation="3" rounded="lg" class="pa-8">
    <v-row cols="12">
      <v-col>
        <v-row justify="center" class="my-4">
          <h3>Add a post</h3>
        </v-row>

        <MessageComponent v-if="message.text" :type="message.type" :text="message.text" />

        <v-form @submit.prevent="handleCreatePost">
          <v-text-field
            id="titleInput"
            v-model="postForm.title"
            label="Title"
            variant="underlined"
            type="text"
            autofocus
          ></v-text-field>
          <v-textarea
            id="descriptionInput"
            v-model="postForm.text"
            label="description"
            variant="underlined"
          ></v-textarea
          ><v-text-field
            id="image_urlInput"
            v-model="postForm.image_url"
            label="image_url"
            variant="underlined"
          ></v-text-field>
          <v-select
            id="categorySelect"
            v-model="postForm.category_id"
            :items="categoriesStore.categories"
            item-title="name"
            item-value="id"
            label="category"
            variant="underlined"
          ></v-select>

          <v-row justify="center" class="mb-4">
            <v-btn id="submitButton" color="red" class="text--white px-8" type="submit">Submit</v-btn>
          </v-row>
        </v-form>
      </v-col>
    </v-row>
  </v-sheet>
</template>

<style scoped lang="scss">
.text--white {
  color: white !important;
}
</style>
