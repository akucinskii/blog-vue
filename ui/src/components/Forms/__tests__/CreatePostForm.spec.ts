import { createTestingPinia } from "@pinia/testing";
import { mount, shallowMount, VueWrapper } from "@vue/test-utils";
import { nextTick } from "vue";
import { mocked } from "ts-jest/utils";
import apiRequest from "../../../utils/apiRequest";
import CreatePostForm from "../CreatePostForm.vue";
import MessageComponent from "../MessageComponent.vue";

let wrapper: VueWrapper<InstanceType<typeof CreatePostForm>>;

jest.mock("../../../utils/apiRequest");
const mockedApiRequest = mocked(apiRequest);

describe("createPostForm", () => {
  beforeEach(() => {
    wrapper = shallowMount(CreatePostForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });
  });

  test("renders without crashing", () => {
    expect(wrapper.exists()).toBe(true);
  });

  test("matches snapshot", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test("form on submit adds a post", async () => {
    const testPostForm = {
      title: "test",
      text: "test",
      category_id: "test",
      image_url: "test",
      big_image_url: "test",
      disabled: false,
    };

    wrapper.vm.postForm = testPostForm;

    await wrapper.find("v-form").trigger("submit.prevent");

    expect(mockedApiRequest).toHaveBeenCalledWith("POST", "/posts", testPostForm);
  });

  test("form on submit displays success message", async () => {
    wrapper = mount(CreatePostForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });

    await wrapper.find("v-form").trigger("submit.prevent");
    await nextTick();

    expect(wrapper.findComponent(MessageComponent).text()).toContain("Post created successfully");
  });

  test("form on submit displays error message", async () => {
    wrapper = mount(CreatePostForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });

    mockedApiRequest.mockRejectedValueOnce({
      response: {
        data: {
          status: 400,
          detail: "test",
        },
      },
    });

    await wrapper.find("v-form").trigger("submit.prevent");
    await nextTick();

    expect(wrapper.findComponent(MessageComponent).text()).toContain("test");
  });

  test("form on submit displays error message when there is no response", async () => {
    wrapper = mount(CreatePostForm, {
      global: {
        plugins: [createTestingPinia()],
      },
    });

    mockedApiRequest.mockRejectedValueOnce({});

    await wrapper.find("v-form").trigger("submit.prevent");
    await nextTick();

    expect(wrapper.findComponent(MessageComponent).text()).toContain("Something went wrong");
  });
});
