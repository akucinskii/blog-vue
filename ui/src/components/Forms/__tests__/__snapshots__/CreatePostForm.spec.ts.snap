// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPostForm matches snapshot 1`] = `
<v-sheet
  class="pa-8"
  elevation="3"
  rounded="lg"
>
  <v-row
    cols="12"
  >
    <v-col>
      <v-row
        class="my-4"
        justify="center"
      >
        <h3>
          Add a post
        </h3>
      </v-row>
      <!--v-if-->
      <v-form>
        <v-text-field
          autofocus=""
          id="titleInput"
          label="Title"
          modelvalue=""
          type="text"
          variant="underlined"
        />
        <v-textarea
          id="descriptionInput"
          label="description"
          modelvalue=""
          variant="underlined"
        />
        <v-text-field
          id="image_urlInput"
          label="image_url"
          modelvalue=""
          variant="underlined"
        />
        <v-select
          id="categorySelect"
          item-title="name"
          item-value="id"
          items=""
          label="category"
          modelvalue=""
          variant="underlined"
        />
        <v-row
          class="mb-4"
          justify="center"
        >
          <v-btn
            class="text--white px-8"
            color="red"
            id="submitButton"
            type="submit"
          >
            Submit
          </v-btn>
        </v-row>
      </v-form>
    </v-col>
  </v-row>
</v-sheet>
`;
